target/riscv/cpu_helper.c

```diff
@@ -905,6 +957,31 @@ static int riscv_cpu_tlb_fill_impl(CPURISCVState *env, vaddr address, int size,
                     TARGET_FMT_plx " prot %d\n",
                     __func__, im_address, ret, *pa, prot2);
 
+#if defined(TARGET_CHERI) && !defined(TARGET_RISCV32)
+            /*
+             * CHERI's load-side caveats are enforced only on the guest
+             * tables at the moment.  Because we are about to AND the two
+             * prot words together, *set* both caveats in prot2 so that
+             * either bit will be preserved from prot.
+             */
+            prot2 |= PAGE_LC_TRAP | PAGE_LC_CLEAR;
+
+            /*
+             * XXX Eventually we probably want to permit the hypervisor to be
+             * able to force tag clearing or trapping.  In decreasing priority
+             * order, if the tag comes back set...
+             *
+             *  - If the hypervisor says trap, raise a hypervisor (HS) fault
+             *
+             *  - If the guest says trap, raise a supervisor (VS/S) fault.
```

Comment:
Other way round, no? Guest translation happens first, and host translation only happens if that succeeds, so the same should be true for tag-dependent traps due to PTEs? Remembering which is going to be annoying though... but Lawrence has probably had to do that for Morello?